[PetrichorPrecipice] iP

PetrichorPrecipice opened the pull request:
# **Alpha**

Alpha is a Chatbot I made with my very limited programming skills. It currently is: 

- Purely text-based
- Probably disappointed at how poorly I made it
- ~~Functional~~ Usable

Here's how I developed Alpha:
1. Bid for CS2103T
2. Appeal for CS2103T
3. Spend like a solid 10 hours weekly constantly changing the code
4. And it is finally **done**. (Until next week :sob:)

Features include:
- [x] These fancy checkbox thingies that you cannot click on. 

You can go [here](https://github.com/nus-cs2103-AY2324S1/ip) to view Alpha's predecessor (even though Alpha should really be the first). 

The code I am proudest of in Alpha:
```
private static String TODO = "todo";
```

And, last but not least:

> "Bye." - Chatbot Alpha

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

PetrichorPrecipice committed:
Change Chat Bot name

PetrichorPrecipice committed:
Added Echo

PetrichorPrecipice committed:
Added List class with add, list functionality

PetrichorPrecipice committed:
Added the checking and unchecking of tasks with new Task class

PetrichorPrecipice committed:
Added Event, Deadline, ToDo classes

PetrichorPrecipice committed:
Added Exception Handling

PetrichorPrecipice committed:
Added delete functionality and Automated UI Testing

PetrichorPrecipice committed:
Added File Storage

PetrichorPrecipice committed:
Add Dates and Times

PetrichorPrecipice committed:
Add more OOP

PetrichorPrecipice committed:
Merge branch 'branch-A-MoreOOP'

# Conflicts:
#	src/main/java/Alpha.java
#	src/main/java/Deadline.java
#	src/main/java/Event.java
#	src/main/java/FileHandler.java

PetrichorPrecipice committed:
Add More OOP

PetrichorPrecipice committed:
Merge branch 'branch-A-MoreOOP'

# Conflicts:
#	src/main/java/Alpha.java

PetrichorPrecipice committed:
Revert "Add More OOP"

This reverts commit 118eb7d9330e4a891fb7ab073409598107b17179.

PetrichorPrecipice committed:
Remove Unused Import

PetrichorPrecipice committed:
Merge branch 'branch-A-MoreOOP' of https://github.com/PetrichorPrecipice/ip

# Conflicts:
#	src/main/java/Alpha.java

PetrichorPrecipice committed:
Merge branches 'master' and 'master' of https://github.com/PetrichorPrecipice/ip

# Conflicts:
#	src/main/java/Alpha.java
#	src/main/java/Deadline.java
#	src/main/java/Event.java
#	src/main/java/FileHandler.java

PetrichorPrecipice committed:
Add More OOP

PetrichorPrecipice committed:
Add Package alpha

PetrichorPrecipice committed:
Merge branch 'add-gradle-support' of https://github.com/PetrichorPrecipice/ip

PetrichorPrecipice committed:
Add JUnit Tests

PetrichorPrecipice committed:
Add JAR file code

PetrichorPrecipice committed:
Add JavaDoc

PetrichorPrecipice committed:
Tweak code to comply with coding standard

PetrichorPrecipice committed:
Add Find functionality

PetrichorPrecipice committed:
Update README.md

PetrichorPrecipice committed:
Replace CLI with GUI

PetrichorPrecipice committed:
Merge branch 'master' of https://github.com/PetrichorPrecipice/ip

DeltonCheng commented on src/main/java/alpha/FileHandler.java:
```diff
@@ -0,0 +1,192 @@
+     */
+    // Checks or unchecks a task in the file. Does so by creating a temp file that copies everything over
+    // except the task that is being checked/unchecked.
+    public void checkOrUncheck(int index, boolean check) {
```
Any reason why you used a temp file instead of just overwriting the current file? Not sure if this will affect the efficiency

DeltonCheng commented on src/main/java/alpha/FileHandler.java:
```diff
@@ -0,0 +1,192 @@
+     * @param index Index of the task to be deleted.
+     */
+    // Deletes a task by creating a temp file and copying everything but the deleted task over.
+    public void delete(int index) {
```
same as above regarding the temp file, and perhaps should use a method name that is more specific

DeltonCheng commented on src/main/java/alpha/Parser.java:
```diff
@@ -0,0 +1,238 @@
+     * @throws MissingIndexException if the index is missing.
+     * @throws InvalidIndexException if the index exceeds or is lesser than the number of stored tasks.
+     */
+    public Command delete(String input) throws MissingIndexException, InvalidIndexException {
```
perhaps a more specific name for the methods, and for the other methods in several other places as well

DeltonCheng commented on README.md:
````diff
@@ -1,24 +1,29 @@
-   | |_| | |_| |   <  __/
-   |____/ \__,_|_|\_\___|
-   ```
+
````
i like that you have updated the README :)

DeltonCheng reviewed (commented):
Overall, code is pretty good. Perhaps could be more specific in terms of naming the methods  and package the related files together (eg. exceptions, commands) for easier reading. Good job!

JasonRay168 commented on src/main/java/alpha/UI.java:
```diff
@@ -42,35 +41,27 @@ public void goodbye() {
-                LINE_SEPARATOR);
+    public String taskAdded(Task task, int size) {
+        return "Alright! I've added this task:\n " + " " + task
+                + "\nNow you have " + size + " tasks in the list.\n";
```
I like how you remembered to add 8 spaces instead of 4 for these cases where the line is broken in the middle

JasonRay168 commented on src/main/java/alpha/UI.java:
```diff
@@ -80,39 +71,29 @@ public void unmark(Task task) {
-                "Noted. I've removed this task:\n" + "  " + task + "\n Now You have " + size +
-                " tasks in the list." + "\n" + LINE_SEPARATOR);
+    public String delete(Task task, int size) {
+        return "Noted. I've removed this task:\n" + "  " + task + "\n Now You have " + size +
+                " tasks in the list.";
```
Maybe can move the '+' sign from the end of a line to the start of the next line? Also, make the split such that the two halves are not too different in length.

So can rewrite as something like this:
```ruby
return "Noted. I've removed this task:\n" + "  " + task
        + "\n Now You have " + size + " tasks in the list.";
```

JasonRay168 commented on src/main/java/alpha/Alpha.java:
```diff
@@ -0,0 +1,144 @@
+        fileHandler.checkAndCreate();
+    }
+
+    /**
+     * Starts the Alpha ChatBot. Ends when "bye" is input.
+     */
+    /*
+    public void run() {
+        ui.introduce();
+        boolean isExit = false;
+        while (!isExit) {
+            String input = ui.read();
+            Command command = parser.parse(input);
+            command.execute();
+            isExit = command.isExit();
+        }
+        ui.goodbye();
+    }
+     */
```
I think it would be better if you add a description on why you commented this code out. If there is no use for this method, then you can delete it entirely to reduce the amount of comments.

JasonRay168 commented on src/main/java/alpha/DeleteCommand.java:
```diff
@@ -0,0 +1,18 @@
+
+    private int index;
+
+    public DeleteCommand(TaskList taskList, FileHandler fileHandler, UI ui, int index) {
+        super(taskList, fileHandler, ui);
+        isExit = false;
+        this.index = index;
+    }
```
Create JavaDoc comments for this and other classes as well, as you have done in Alpha.java and Command.java

JasonRay168 commented on src/main/java/alpha/Parser.java:
```diff
@@ -0,0 +1,238 @@
+                return delete(input);
+            } else if (splitInput[0].equals(FIND)) {
+                return find(input);
+            } else {
+                throw new InvalidInputException("Invalid Input!");
+            }
+        } catch (MissingIndexException e1) {
+            return new InvalidCommand(taskList, fileHandler, ui, e1.getMessage() +
+                    " Please enter the index of the number you would like to mark.");
+        } catch (InvalidIndexException e2) {
+            return new InvalidCommand(taskList, fileHandler, ui, e2.getMessage() +
+                    " Please enter a valid index. To check all valid indices, " +
+                    "type \"list\" and press ENTER");
+        } catch (InvalidInputException e3) {
+            return new InvalidCommand(taskList, fileHandler, ui, e3.getMessage()
+                    + " Please input something meaningful.");
+        } catch (NumberFormatException e4) {
+            return new InvalidCommand(taskList, fileHandler, ui, "Please enter a number.");
+        } catch (InvalidFormatException | MissingInfoException e5) {
+            if (e5.getTask() == TaskException.TaskType.TODO) {
+                return new InvalidCommand(taskList, fileHandler, ui, e5.getMessage()
+                        + " Please enter a todo in the " +
+                        "format \"todo YOUR_DESCRIPTION\"");
+            } else if (e5.getTask() == TaskException.TaskType.DEADLINE) {
+                return new InvalidCommand(taskList, fileHandler, ui, e5.getMessage()
+                        + " Please enter a deadline in the format " +
+                        "\"deadline YOUR_DESCRIPTION /by YOUR_TIME\" ");
+            } else if (e5.getTask() == TaskException.TaskType.EVENT) {
+                return new InvalidCommand(taskList, fileHandler, ui,
+                        " Please enter an event in the format " +
+                        "\"event YOUR_DESCRIPTION " +
+                        "/from START_TIME /to END_TIME\" ");
+            }
+            return new InvalidCommand(taskList, fileHandler, ui, "");
+        } catch (DateTimeParseException e6) {
+            return new InvalidCommand(taskList, fileHandler, ui,
+                    e6.getMessage() +"The date is in an invalid format! " +
+                            "Enter the date in the format YYYY-MM-DD");
+        }
+    }
```
This method is quite long, maybe instead of using if else, you can use switch case to reduce space. Alternatively can remove some of the new lines / enters in the lines so it would not consume too many lines

JasonRay168 reviewed (commented):
I think your PR code is relatively clean and easy to navigate, just a few minor coding standard issues you can work on.

PetrichorPrecipice committed:
Add assert statements

Code in classes like "Parser" and "FileHandler" lack assertions to
document important assumptions about input data and behavior.

The lack of assertions can hamper code readability and cause some
unexpected behaviours to go unnoticed.

Let's add assertions to various methods in "Parser" and "FileHandler"
classes to ensure the expected conditions and constraints are met within
each method.

Adding assertions clarifies these assertions and validates assumptions.
They help enhance code readability and maintainability.

PetrichorPrecipice committed:
Handle Duplicate Tasks Entered

Possible to enter two same tasks with same description, date and time.

This does not make sense as the user should not have two of the same
tasks. They may have either forgotten they have entered the same task
before or mistyped.

Let's handle exceptions so that every time the user tries to add a task,
the Parser class checks if that task is already in the taskList through
an overriden equals method for each of the subclasses of Task. If a task
is in the task list, do not add it and inform the user.

Overriding the equals method is preferable as it is more maintainable
and extensible compared to hard-coding the equality check into the
method that checks against every task in the task list.

PetrichorPrecipice committed:
Refactor code quality

All classes are in the same "alpha" package.

Problematic because there is no clear distinction between classes,
all classes are just grouped together.

Separated the classes into 4 packages that are organised more clearly
based on their functions. Also refactored code in other places to
conform to coding quality standards.

Clearer organisation of packages improves code maintainability and
readability.

PetrichorPrecipice committed:
Remove unnecessary code

PetrichorPrecipice committed:
Merge pull request #2 from PetrichorPrecipice/branch-A-CodeQuality

Remove unnecessary code

PetrichorPrecipice committed:
Change UI elements

PetrichorPrecipice committed:
Merge pull request #3 from PetrichorPrecipice/branch-A-CodeQuality

Change UI elements

PetrichorPrecipice committed:
Update Font

PetrichorPrecipice committed:
Merge pull request #4 from PetrichorPrecipice/branch-A-BetterGui

Update Font

PetrichorPrecipice committed:
Improve Code Quality

PetrichorPrecipice committed:
Update UG

PetrichorPrecipice committed:
Add Screenshot

PetrichorPrecipice committed:
Update Screenshot

PetrichorPrecipice committed:
Update User Guide

PetrichorPrecipice committed:
Various Bug Fixes and Update UG

PetrichorPrecipice committed:
Update UG

PetrichorPrecipice committed:
Various Bug Fixes
